File: c++/hand.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include "hand.h"
#include "card.h"

using namespace std;

const vector<string> Hand::HAND_MAP = { "High Card", "Pair", "Two Pair", "Three of a Kind", "Straight",
                "Flush", "Full House", "Four of a Kind", "Straight Flush", "Royal Straight Flush" };

Hand::Hand() {}

ostream& operator<<(ostream& os, const Hand& h) {
    for (int i = 0; i < h.cards.size(); i++)
        os << h.cards[i];

    return os;
}

bool operator>(const Hand& hand1, const Hand& hand2) {
    return hand1.getType() > hand2.getType();
}

//=============== Public Methods ===============
void Hand::addCard(Card c) {
    cards.push_back(c);
}

Hand::HAND_TYPE Hand::getType() const {
    return type;
}

void Hand::setType(Hand::HAND_TYPE t) {
    type = t;
}

vector<Card> Hand::getSortedCards() const {
    vector<Card> sortedCards(cards);
    sort(sortedCards.begin(), sortedCards.end());

    return sortedCards;
}

File: julia/hand.jl
const HAND_SIZE = 5
const HAND_MAP =  ["High Card", "Pair", "Two Pair", "Three of a Kind", "Straight",
                "Flush", "Full House", "Four of a Kind", "Straight Flush", "Royal Straight Flush"]

mutable struct Hand
    cards :: Vector{Card}
    type :: Int64
end

function Hand()
    hand = Hand(Vector{Card}(undef, 0), -1)
    return hand
end

function add_card(hand, card)
    push!(hand.cards, card)
end

function get_sorted_cards(hand::Hand)
    return sort(hand.cards)
end

function set_type(hand::Hand, type::Int64)
    hand.type = type
end

function Base.show(io::IO, hand::Hand)
    for card in hand.cards
        print(card)
    end
end
